### ISSUE TITLE
[ActiveRecord] Issue with joins && merges

### ISSUE THREAD

**Issue by gustavolobo** [rails/rails#33379]
### Steps to reproduce

```ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  gem 'activerecord', '5.2.0'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :users, force: true  do |t|
    t.string :name
    t.boolean :locked
  end

  create_table :blogger_profiles, force: true  do |t|
    t.integer :user_id
    t.boolean :verified
  end
end

class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true
end

class User < ApplicationRecord
  has_many :blogger_profiles

  scope :not_locked, -> { where(locked: false) }
  scope :verified, -> { joins(:blogger_profiles).merge(BloggerProfile.verified).distinct }
end

class BloggerProfile < ApplicationRecord
  belongs_to :user

  scope :verified, -> { joins(:user).where(verified: true).merge(User.not_locked) }
end

class BugTest < Minitest::Test
  def test_query_stuff
    User.verified.to_a
  end
end
```

### Expected behavior
To not raise an exception.

### Actual behavior
Raises an exception in 5.2.0 (`ActiveRecord::StatementInvalid: SQLite3::SQLException: ambiguous column name: main.users.id`), it works in 5.1.6.

The query generated in 5.1.6:
```sql
SELECT DISTINCT "users".* FROM "users"
INNER JOIN "blogger_profiles" ON "blogger_profiles"."user_id" = "users"."id" 
LEFT OUTER JOIN "users" "users_blogger_profiles" ON "users_blogger_profiles"."id" = "blogger_profiles"."user_id" 
WHERE "blogger_profiles"."verified" = ? AND "users"."locked" = ?
```

The query generated in 5.2.0:
```sql
SELECT DISTINCT "users".* FROM "users" 
INNER JOIN "blogger_profiles" ON "blogger_profiles"."user_id" = "users"."id" 
INNER JOIN "users" ON "users"."id" = "blogger_profiles"."user_id" 
WHERE "blogger_profiles"."verified" = ? AND "users"."locked" = ?
```

**Comment by twitnithegirl** [comment]
I was able to reproduce this against 5.2.0 and current master.

**Comment by gustavolobo** [comment]
What's the process to get this fixed?
Happy to help if anyone point me in the right direction.

**Comment by twitnithegirl** [comment]
@gustavolobo you can [fork the repo, fix the bug, and submit a PR with your patch](http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-code) however if you aren't able to do that I can probably start investigating it tomorrow and see what I can come up with.

**Comment by kamipo** [comment]
This should be fixed by #33066.

**Comment by rafaelfranca** [comment]
@kamipo should we backport?

**Comment by kamipo** [comment]
had been backported in 6869c5d02fd1478ad87977b658d9b4c2b17121ed.

**Comment by rafaelfranca** [comment]
Weird. It was no linked in the PR. thanks

**Comment by gustavolobo** [comment]
@rafaelfranca @kamipo awesome news.
I don't see this included in the milestone for 5.2.1 though. 
What's the release process? When can we expect this released?
In the meantime, we can use the git reference or the branch 5.2-stable. Just checking.

**Comment by aleksandrs-ledovskis** [comment]
For any future reference, this was fixed in v5.2.1 albeit in reworked form via https://github.com/rails/rails/commit/50036e673b3b6ff41476814b212b55fabac3a638
